libraries/chain/escrow_evaluator.cpp

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2018 oxarbitrage and contributors.
+ *
+ * The MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include <graphene/chain/database.hpp>
+#include <graphene/chain/escrow_evaluator.hpp>
+#include <graphene/chain/escrow_object.hpp>
+#include <graphene/chain/hardfork.hpp>
+
+namespace graphene { namespace chain {
+
+      void_result escrow_transfer_evaluator::do_evaluate(const escrow_transfer_operation& o)
+      {
+         FC_ASSERT( db().head_block_time() > HARDFORK_ESCROW_TIME,
+                    "Operation not allowed before HARDFORK_ESCROW_TIME."); // remove after HARDFORK_ESCROW_TIME
+
+         FC_ASSERT( o.ratification_deadline > db().head_block_time() );
+         FC_ASSERT( o.escrow_expiration > db().head_block_time() );
+         FC_ASSERT( db().get_balance( o.from, o.amount.asset_id ) >= (o.amount + o.fee + o.agent_fee) );
```

Comment:
What if amount, fee and agent_fee have different asset_id?

Reply:
those checks are done in `validate()` at `escrow.cpp` however the fee==bts was not enforced. added here https://github.com/bitshares/bitshares-core/pull/866/commits/cb779d1175bdbdc9a70bdd538183281295e6bab4 this is temporal until we decide we an make the escrow works with other amount than bts then some of this rules will probably change.